Simplify EarlyStopper

mberr opened the pull request:
This PR simplifies the EarlyStopper implementation, and adds a few smaller features:

* [x] Change evaluation epochs from `1, (frequency + 1), (2 * frequency + 1), ...` to `frequency, (2 * frequency), ...`
* [x] Save epoch at which we had the best result
* [x] Export best epoch and best result via `get_summary_dict`
* [x] Replace ring buffer by `remaining_patience` counter, which gets decreased every time we evaluate and do not observe an improvement. If we observe such, the counter is reset to `patience`.
* [x] Change detection of significant improvement from using an absolute delta to a relative one (which should be easier to set).

Depends on #72

mberr committed:
Simplify EarlyStopper

cthoyt committed:
Remove unused import

mberr committed:
Merge remote-tracking branch 'origin/master' into simplify-early-stopper

# Conflicts:
#	src/pykeen/stoppers/early_stopping.py
#	tests/test_early_stopping.py

mberr committed:
Post merge fix

mberr committed:
Remove resolved TODO

[skip ci]

mberr committed:
Save best epoch

[skip ci]

mberr committed:
Add best epoch and result to get_summary_dict

mberr committed:
Fix indention

mberr commented on tests/test_early_stopping.py:
```diff
@@ -235,7 +240,7 @@ class TestEarlyStoppingRealWorld(unittest.TestCase):
     max_num_epochs: int = 1000
     #: The epoch at which the stop should happen. Depends on the choice of random seed.
-    stop_epoch: int = 21
+    stop_epoch: int = 30
```
Changing the evaluation times is a somewhat breaking change, in the sense that when training exactly the same model with fixed random seed and early stopping than we did before this PR it will stop at a different epoch (which is why the `stop_epoch` changed). However, I would argue that evaluating at epoch `frequency`, `2* frequency` is more natural and what you would expect when setting an evaluation frequency, rather than `1`, `frequency + 1`, `2 * frequency + 1`, etc.

cthoyt replied:
this breaking change is fine with me. I think the new way makes  more sense

cthoyt commented on src/pykeen/stoppers/early_stopping.py:
```diff
@@ -17,52 +15,43 @@
-    return result > baseline
-
+    if larger_is_better:
+        best_value = -best_value
```
I would prefer a different return here that is 

```python
return -current_value < (-best_value - absolute_delta)
```

Or simplified

```python
return current_value > best_value + absolute_delta
```

mberr replied:
1664e2d 

I also changed from using an `absolute_delta` to a relative one, which should be easier to set.

cthoyt commented on src/pykeen/stoppers/early_stopping.py:
```diff
@@ -116,20 +107,15 @@ def __post_init__(self):
     def should_evaluate(self, epoch: int) -> bool:
         """Decide if evaluation should be done based on the current epoch and the internal frequency."""
-        return 0 == ((epoch - 1) % self.frequency)
+        return epoch > 0 and epoch % self.frequency == 0
```
looking back, it the old code is really crazy. I don't know why I did it like this

cthoyt commented on src/pykeen/stoppers/early_stopping.py:
```diff
@@ -157,22 +143,30 @@ def should_stop(self, epoch: int) -> bool:
-
         # Append to history
         self.results.append(result)
+        self.number_evaluations += 1
```
can't number_evaluations just be calculated as `len(self.results)`? Maybe also set as a property? Or do you prefer to keep it explicit so later it gets dumped with the dataclass?

mberr replied:
There is already an `number_results` exactly doing this :see_no_evil: I am going to replace `number_evaluations` by that.

mberr replied:
e443d46

cthoyt committed:
Fix type annotation

cthoyt committed:
Add result callbacks

This allowed a bit of simplification of the HPO callbacks, and also makes it a bit more flexible/extensible

cthoyt reviewed (approved):
@mberr note I made a change in the way the callbacks work - it should work the same way but afford a bit more flexibility for anyone who wants to muck around in it themselves

mberr committed:
Add relative delta for minimum relative improvement

mberr committed:
Change default minimum change from absolute to relative

mberr committed:
Replace manual counter number_evaluations by existing property number_results

mberr committed:
Improve info message when early stopping

mberr committed:
Fix early stopping tests usage of delta

mberr committed:
Merge remote-tracking branch 'origin/simplify-early-stopper' into simplify-early-stopper

mali-git commented on src/pykeen/stoppers/early_stopping.py:
```diff
@@ -17,52 +15,45 @@
+    best_value: float,
+    current_value: float,
+    larger_is_better: bool,
+    absolute_delta: float = 0.0,
```
Is `absolute_delta` defined somewehere else, too (it seems that `absolute_delta` is never explicitly passed to the function)?

mberr replied:
No, since I changed the default from using absolute delta for significant improvement to relative delta. I was not sure whether we should expose both via the EarlyStopper. Any suggestions for that? Otherwise, we can remove this parameter.

mali-git replied:
For keeping the `EarlyStopper` simple, I think it is better to decide for one of those.

mberr replied:
I removed the `absolute_delta` variant 
d5742d2

mali-git replied:
Thanks!

mberr committed:
Remove absolute_delta

mberr committed:
Adjust key of summary_dict
